pkp/pkp-lib#2906 Add EmailTemplates API and ListPanel

NateWr opened the pull request.

NateWr committed:
pkp/pkp-lib#2906 Initial refactor of ListPanels and Email Templates

- Add /emailTemplates endpoint and Service class.
- Integrate new ListPanel containers.

NateWr committed:
pkp/pkp-lib#2906 Run email template upgrades before schema

NateWr committed:
pkp/pkp-lib#2906 Add restoreDefaults endpoint to email templates API docs

NateWr committed:
pkp/pkp-lib#2906 Complete integration of EmailTemplatesListPanel

- Use common locale string for series editors
- Use new SettingsContainer component
- Update role ids in email template defaults
- Fix series editor tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pkp/omp?pullRequest=657) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pkp/omp?pullRequest=657) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pkp/omp?pullRequest=657) it.</sub>

NateWr committed:
pkp/pkp-lib#2906 Remove merge conflict code introduced during rebase

NateWr committed:
pkp/pkp-lib#2906 Fix fatal error calling Application from within namespace

NateWr committed:
pkp/pkp-lib#2906 Fix series editor selection test

NateWr committed:
pkp/pkp-lib#2906 Fix series editor tests

NateWr committed:
pkp/pkp-lib#2906 Try to resolve occassional test failures with send to review

asmecher commented on dbscripts/xml/upgrade.xml:
```diff
@@ -41,6 +41,10 @@
 		<data file="dbscripts/xml/upgrade/3.1.2_preupdate_pub_id.xml" condition="return $installer->columnExists('published_submissions', 'pub_id');" />
 	</upgrade>
 
+	<upgrade minversion="0.9.9.0" maxversion="3.1.2.9">
```
This should probably be `maxversion="3.1.9.9"`.

asmecher commented on dbscripts/xml/upgrade/3.2.0_preupdate_email_templates.xml:
```diff
@@ -0,0 +1,94 @@
+		<query>
+			CREATE TABLE email_templates_settings (
+				email_id INT,
+				locale VARCHAR(14) NOT NULL DEFAULT "",
```
Quotes should be `''` (not `""`) for PostgreSQL

asmecher commented on dbscripts/xml/upgrade/3.2.0_preupdate_email_templates.xml:
```diff
@@ -0,0 +1,94 @@
+			)
+		</query>
+		<query>
+			CREATE TABLE email_templates_settings (
```
The table creation should be handled by the schema XML descriptors -- if those are listed in `dbscripts/xml/upgrade.php` then there's no need to include `CREATE TABLE` statements.

NateWr replied:
In this case, I'm doing a little bit of extra work and I wasn't sure if it was possible to do it purely through the schema. Here's a quick rundown of what's happening:

1. Create `email_id` column in `email_templates_data`.
2. Populate that column with values from `email_templates`.
3. Use the new rows in `email_templates_data` to feed a new table, `email_templates_settings`.
4. Delete the old `email_templates_data` table.

So in this case, I can't rely on the schema to just drop and create tables, because I'm moving data over from one to the other. Is there a better way to do this?

The same is true for the `context_id` column added to `email_templates`. I add this column, then populate it with the values based on two other columns, before dropping the two other columns.

asmecher commented on dbscripts/xml/upgrade/3.2.0_preupdate_email_templates.xml:
```diff
@@ -0,0 +1,94 @@
+			DROP TABLE email_templates_data
+		</query>
+		<query>
+			ALTER TABLE email_templates
```
Likewise, new columns should be left to the schema descriptors (and column changes).

asmecher commented on dbscripts/xml/upgrade/3.2.0_preupdate_email_templates.xml:
```diff
@@ -0,0 +1,94 @@
+		<query>
+			UPDATE email_templates_default
+			SET from_role_id=4097
+			WHERE from_role_id=768 OR from_role_id=8192
```
Make sure to include commented constant names when hard-coding magic numbers. (Eases grepping.)

asmecher reviewed (changes requested):
Just a few changes!

NateWr committed:
pkp/pkp-lib#2906 Fixes after code review

- Update copyright date
- Use better maxversion in upgrade scripts
- Fix space-based indents
- Use single quotes for postgresql compatibility
- Add constant names to upgrade scripts

NateWr committed:
pkp/pkp-lib#2906 Update tests to use button instead of link

asmecher reviewed (changes requested):
On the migration scripts:

If you need an autoincrement column added to an existing table, you'll probably blow ADODB's tiny mind (and rightly so, as MySQL and PostgreSQL see autoincrements very differently). We've used a pattern in a few places that could help:
1. Rename the existing `email_templates_data` table to `email_templates_data_migration` (see e.g. [this rename statement](https://github.com/pkp/ojs/blob/ojs-stable-3_1_0/dbscripts/xml/upgrade/3.0.0_preupdate.xml#L50)). This runs before the schema descriptors are synced.
2. Let the schema descriptor recreate the `email_templates_data` (now with the new ID column).
3. Use an `INSERT INTO ... SELECT` statement to repopulate `email_templates_data` from `email_templates_data_migration` and whatever else. (See e.g. [this one](https://github.com/pkp/ojs/blob/ojs-stable-3_1_0/dbscripts/xml/upgrade/3.0.0_update.xml#L175).)
4. Drop the `email_templates_migration` table when we're done. (See e.g. [this](https://github.com/pkp/ojs/blob/ojs-stable-3_1_0/dbscripts/xml/upgrade/3.0.0_postupdate.xml#L16).)

Would that work?

NateWr commented:
Apologies if I've misunderstood you, but I'm not actually adding an auto increment column to a table. The first move adds an `email_id`, but it's not auto increment and the whole `email_templates_data` table will be removed (and not recreated by schema):

```
ALTER TABLE email_templates_data ADD email_id BIGINT NOT NULL
```

Basically, all I am doing is prepping for the next statement, which links each row in `email_templates_data` with a row in `email_templates`:

```
UPDATE email_templates_data as etd
SET etd.email_id = (
  SELECT et.email_id
  FROM email_templates as et
  WHERE etd.email_key = et.email_key
    AND etd.assoc_type = et.assoc_type
    AND etd.assoc_id = et.assoc_id
)
```

And all that I'm doing there is prepping for these statements, which split out each row in `email_templates_data` into the `<entity>_settings` table schema that is common to other entities:

```
INSERT INTO email_templates_settings (email_id, locale, setting_value, setting_name)
SELECT email_id, locale, subject, 'subject'
FROM email_templates_data
INSERT INTO email_templates_settings (email_id, locale, setting_value, setting_name)
SELECT email_id, locale, body, 'body'
FROM email_templates_data
```

Then I drop `email_templates_data` completely. No auto increment columns have been modified. I've simply moved the data into an `<entity>_settings` table format.

I can rename `email_templates_data` to `email_templates_data_migration` before doing this, but I will still have to juggle the data in the same way. And the schema will not recreate the table or anything for me, since it's no longer in the schema. Am I missing the point?

asmecher commented:
Gotcha, Nate -- we've never done `ALTER TABLE` in the upgrade scripts explicitly before, but since it's simple (e.g. type names are agreed on by both MySQL and PostgreSQL) and the table is being put to death anyway, I think it's OK to do that way. However, you'll need to leave the `NOT NULL` constraint off the `ALTER TABLE ADD COLUMN`, since it doesn't have a default.

NateWr commented:
Ok thanks @asmecher! I've removed the `NOT NULL` and the tests are running. They probably won't finish before I'm off today.

If you see them pass, you can merge this. Otherwise I'll follow-up when I'm back in a week.

asmecher commented:
Unfortunately the tests didn't make it -- have a good trip and we'll work on this afterwards!

NateWr committed:
Submodule update ##NateWr/i2906_email_templates##
